docs/design/query-caching.md

```diff
@@ -0,0 +1,224 @@
+
+The size of the value stored is variable and dependent on the data returned by the database. 
+
+#### Size
+
+The maximum cache size is dependent on the memory (RAM) specifications of the hosting server. The cache will share memory space with the running instance of DAB and other operating system/hosting resources. 
+Since key/value pairs are not consistently sized, we can utilize string length (character count) to determine a relative size of the cache entry. 
+
+### Cache Behavior
+
+#### Eviction
+
+Cache entries are evicted when the cache gets close to our predetermined size limit, and when entries expire based on their configured ttl.
+
+#### Updates
+
+Detect if the operation is a follow up read to a mutation. We can then map that follow up read generated query to a cache key entry and attempt to invalidate the cache entry. This would not allow us to expire all variations of read queries generated for the specific entity(s) because querystrings, role based access, and column based access may be cached with different keys. This behavior backs up our guidance to limit caching to rarely changing database objects/records because cache entries may be stale and fed back to the user.
+
+## Security and privacy
+
+### Data residency
+
+__Cache data is stored in-memory__ and remains in memory until evicted due to ttl rules, manual eviction, or DAB process shutdown. Cache data is not persisted to disk.
+The cache will store many representations of database data due to the variations of SQL queries generated and executed against the database. Some queries may include or exclude certain database columns depending on the role of the requestor. 
+
+__Authorization configuration does not explicitly affect how the caching service operates__ because all authorization actions execute prior cache interaction. Requests will be rejected due to authorization restrictions prior to cache interaction.
+
+__DAB will return unexpired cache entries when a database goes offline.__ Depending on the developer configured ttl for an entity, the cache entries will persist in an offline database scenario for the ttl defined duration. A cache entry is evicted after the ttl is surpassed.
+
+### Data in transit
+
+__Data returned by the database is received over the same TLS connection established to execute queries.__ The caching service described in this document is an in-memory cache and not a distributed cache. Consequently, the cache does not further transmit data returned by the database, by design. 
+
+### Access token metadata usage
+
+__Cache entries will not be created for database query text generated from database policies.__ This is because database policies directly affect values set within SQL Server's `-set-session-context` variables and is an implementation detail of database specific query executors. The caching service will be implemented independent of and before execution of query executor code. 
```

Comment:
If the query executor code decides whether to add database policies metadata in the query corresponding to a given request, how would we identify the request to be a cache miss if a similar request with NO database policies was already cached? 

Would that lead to incorrect cache hits preventing addition of additional predicates based on policies which would have otherwise happened if the cache were to be populated AFTER evaluation of policies?

Basically at which point do we check the cache for a hit or miss? - after complete generation of the query text OR right before executing the query?

Reply:
Good point, I'll need to add logic which knows to skip cache for db policy affected requests. 

Your point also leads me to observe that the database policy/set session context query text generation should not be in the query executor. Ideally, that logic should be in the query /mutation engines or the sql query builders. The Query Executor should have a single purpose of executing the request metadata provided to it and not attempt to further modify the request metadata sent to it.

Reply:
I vaguely remember @ayush3797 ran into some issues with adding session context query text to the query builders, @ayush3797 could you please provide reasoning behind why db policy query text generation was added to the executor instead of the builder?

Reply:
context is: https://github.com/Azure/data-api-builder/pull/1084#discussion_r1066278125

> Creating the query earlier won't simplify things - the sqltext is specific to the request intent (PUT/PATCH/INSERT) etc but the session context query is constant for all of them, so instead of separately adding it to the query at distributed places, we can do it at one single place. Also, I'll keep the name of the dictionary to something more generic like sessionParameters because its a good possibility we will have a similar functionality for postgres as well, not sure of MySql.

My interpretation is that there isn't a technical limitation to moving. Result would be more places in code which add the session parameter sql text. To me, that is a valid tradeoff for separation of concerns. During a refactor, we'd be able to better determine how much code de-duplication can occur.

Reply:
Also, session_context is all or nothing. all queries will get session context containing token claims when runtimeconfig's mssqloptions.SetSessionContext == true. so this is another boolean that will gate usage of the cache.